In [ ]:
import pandas as pd
import numpy as np
import datetime as dt

In [ ]:
df = pd.read_csv ('Online Retail Data.csv')

In [ ]:
df

order_id product_code                     product_name  quantity  \
0        493410      TEST001          This is a test product.         5   
1       C493411        21539          RETRO SPOTS BUTTER DISH        -1   
2        493412      TEST001          This is a test product.         5   
3        493413        21724  PANDA AND BUNNIES STICKER SHEET         1   
4        493413        84578   ELEPHANT TOY WITH BLUE T-SHIRT         1   
...         ...          ...                              ...       ...   
461768   539991        21618   4 WILDFLOWER BOTANICAL CANDLES         1   
461769   539991        72741            GRAND CHOCOLATECANDLE         4   
461770   539992        21470    FLOWER VINE RAFFIA FOOD COVER         1   
461771   539992        22258          FELT FARM ANIMAL RABBIT         1   
461772   539992        21155            RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
3       2010-01-04 09:54:00   0.85          NaN  
4       2010-01-04 09:54:00   3.75          NaN  
...                     ...    ...          ...  
461768  2010-12-23 16:49:00   1.25          NaN  
461769  2010-12-23 16:49:00   1.45          NaN  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 7 columns]

In [ ]:
df['quantity'].fillna(df['quantity'].mean())

0         5
1        -1
2         5
3         1
4         1
         ..
461768    1
461769    4
461770    1
461771    1
461772    1
Name: quantity, Length: 461773, dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      461773 non-null  object 
 1   product_code  461773 non-null  object 
 2   product_name  459055 non-null  object 
 3   quantity      461773 non-null  int64  
 4   order_date    461773 non-null  object 
 5   price         461773 non-null  float64
 6   customer_id   360853 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.7+ MB


In [ ]:
df[df['order_id']=='539992']

order_id product_code                   product_name  quantity  \
461770   539992        21470  FLOWER VINE RAFFIA FOOD COVER         1   
461771   539992        22258        FELT FARM ANIMAL RABBIT         1   
461772   539992        21155          RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN

In [ ]:
df[(df['product_name'].str.contains('TEA') | df['product_name'].str.contains('COFFE')) & (df['quantity']>10)]

order_id product_code                       product_name  quantity  \
10       493414        21527      RETRO SPOT TRADITIONAL TEAPOT        12   
22       493427        79000                 MOROCCAN TEA GLASS        12   
112      493433        84991        60 TEATIME FAIRY CAKE CASES        24   
305      493443        21527      RETRO SPOT TRADITIONAL TEAPOT        12   
455      493467        84991        60 TEATIME FAIRY CAKE CASES        24   
...         ...          ...                                ...       ...   
459159   539654        84946    ANTIQUE SILVER TEA GLASS ETCHED        18   
459663   539728        22348        TEA BAG PLATE RED RETROSPOT        12   
460395   539750        22114  HOT WATER BOTTLE TEA AND SYMPATHY       210   
460927   539920        84946    ANTIQUE SILVER TEA GLASS ETCHED        12   
460978   539933        84991        60 TEATIME FAIRY CAKE CASES        24   

                 order_date  price  customer_id  
10      2010-01-04 10:28:00   6.95      14590.0  
22      2010-01-04 10:43:00   0.85      13287.0  
112     2010-01-04 12:51:00   0.55      14709.0  
305     2010-01-04 13:39:00   6.95      14527.0  
455     2010-01-04 14:29:00   0.55      13004.0  
...                     ...    ...          ...  
459159  2010-12-20 17:05:00   1.25      17340.0  
459663  2010-12-21 14:26:00   0.85      16270.0  
460395  2010-12-21 15:40:00   9.32          NaN  
460927  2010-12-23 11:06:00   1.25      14085.0  
460978  2010-12-23 11:24:00   0.55      15235.0  

[4365 rows x 7 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      461773 non-null  object 
 1   product_code  461773 non-null  object 
 2   product_name  459055 non-null  object 
 3   quantity      461773 non-null  int64  
 4   order_date    461773 non-null  object 
 5   price         461773 non-null  float64
 6   customer_id   360853 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.7+ MB


In [ ]:
df.describe()

quantity          price    customer_id
count  461773.000000  461773.000000  360853.000000
mean        9.088892       4.569963   15557.611138
std        86.366467     153.411566    1593.728741
min     -9600.000000  -53594.360000   12346.000000
25%         1.000000       1.250000   14210.000000
50%         3.000000       2.100000   15580.000000
75%        10.000000       4.210000   16938.000000
max     10200.000000   25111.090000   18287.000000

In [ ]:
df.dropna()

order_id product_code                       product_name  quantity  \
0        493410      TEST001            This is a test product.         5   
1       C493411        21539            RETRO SPOTS BUTTER DISH        -1   
2        493412      TEST001            This is a test product.         5   
6        493414        21844                     RETRO SPOT MUG        36   
7        493414        21533          RETRO SPOT LARGE MILK JUG        12   
...         ...          ...                                ...       ...   
461740   539988        84380  SET OF 3 BUTTERFLY COOKIE CUTTERS         1   
461741   539988       84849D              HOT BATHS SOAP HOLDER         1   
461742   539988       84849B             FAIRY SOAP SOAP HOLDER         1   
461743   539988        22854        CREAM SWEETHEART EGG HOLDER         2   
461744   539988       47559B                TEA TIME OVEN GLOVE         2   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
6       2010-01-04 10:28:00   2.55      14590.0  
7       2010-01-04 10:28:00   4.25      14590.0  
...                     ...    ...          ...  
461740  2010-12-23 16:06:00   1.25      18116.0  
461741  2010-12-23 16:06:00   1.69      18116.0  
461742  2010-12-23 16:06:00   1.69      18116.0  
461743  2010-12-23 16:06:00   4.95      18116.0  
461744  2010-12-23 16:06:00   1.25      18116.0  

[360853 rows x 7 columns]

In [ ]:
df.drop(columns='order_id')

product_code                     product_name  quantity  \
0           TEST001          This is a test product.         5   
1             21539          RETRO SPOTS BUTTER DISH        -1   
2           TEST001          This is a test product.         5   
3             21724  PANDA AND BUNNIES STICKER SHEET         1   
4             84578   ELEPHANT TOY WITH BLUE T-SHIRT         1   
...             ...                              ...       ...   
461768        21618   4 WILDFLOWER BOTANICAL CANDLES         1   
461769        72741            GRAND CHOCOLATECANDLE         4   
461770        21470    FLOWER VINE RAFFIA FOOD COVER         1   
461771        22258          FELT FARM ANIMAL RABBIT         1   
461772        21155            RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
3       2010-01-04 09:54:00   0.85          NaN  
4       2010-01-04 09:54:00   3.75          NaN  
...                     ...    ...          ...  
461768  2010-12-23 16:49:00   1.25          NaN  
461769  2010-12-23 16:49:00   1.45          NaN  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 6 columns]

In [ ]:
df_clean = df.copy()
# membuat kolom date
df_clean['date'] = pd.to_datetime(df_clean['order_date']).dt.date.astype('datetime64')
# menghapus semua baris tanpa customer_id
df_clean = df_clean[~df_clean['customer_id'].isna()]
# menghapus semua baris tanpa product_name
df_clean = df_clean[~df_clean['product_name'].isna()]
# membuat semua product_name berhuruf kecil
df_clean['product_name'] = df_clean['product_name'].str.lower()
# menghapus semua baris dengan product_code atau product_name test
df_clean = df_clean[(~df_clean['product_code'].str.lower().str.contains('test')) |
                    (~df_clean['product_name'].str.contains('test '))]
# membuat kolom order_status dengan nilai 'cancelled' jika order_id diawali dengan huruf 'c' dan 'delivered' jika order_id tanpa awalan huruf 'c'
df_clean['order_status'] = np.where(df_clean['order_id'].str[:1]=='C', 'cancelled', 'delivered')
# mengubah nilai quantity yang negatif menjadi positif karena nilai negatif tersebut hanya menandakan order tersebut cancelled
df_clean['quantity'] = df_clean['quantity'].abs()
# menghapus baris dengan price bernilai negatif
df_clean = df_clean[df_clean['price']>0]
# membuat nilai amount, yaitu perkalian antara quantity dan price
df_clean['amount'] = df_clean['quantity'] * df_clean['price']
# mengganti product_name dari product_code yang memiliki beberapa product_name dengan salah satu product_name-nya yang paling sering muncul
most_freq_product_name = df_clean.groupby(['product_code','product_name'], as_index=False).agg(order_cnt=('order_id','nunique')).sort_values(['product_code','order_cnt'], ascending=[True,False])
most_freq_product_name['rank'] = most_freq_product_name.groupby('product_code')['order_cnt'].rank(method='first', ascending=False)
most_freq_product_name = most_freq_product_name[most_freq_product_name['rank']==1].drop(columns=['order_cnt','rank'])
df_clean = df_clean.merge(most_freq_product_name.rename(columns={'product_name':'most_freq_product_name'}), how='left', on='product_code')
df_clean['product_name'] = df_clean['most_freq_product_name']
df_clean = df_clean.drop(columns='most_freq_product_name')
# mengkonversi customer_id menjadi string
df_clean['customer_id'] = df_clean['customer_id'].astype(str)
# menghapus outlier
from scipy import stats
df_clean = df_clean[(np.abs(stats.zscore(df_clean[['quantity','amount']]))<3).all(axis=1)]
df_clean = df_clean.reset_index(drop=True)
df_clean

<ipython-input-26-9b6614ea7552>:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['amount'] = df_clean['quantity'] * df_clean['price']


order_id product_code                       product_name  quantity  \
0       C493411        21539          red retrospot butter dish         1   
1        493414        21844                  red retrospot mug        36   
2        493414        21533          retro spot large milk jug        12   
3        493414        37508    new england ceramic cake server         2   
4        493414       35001G               hand open shape gold         2   
...         ...          ...                                ...       ...   
358464   539988        84380  set of 3 butterfly cookie cutters         1   
358465   539988       84849D              hot baths soap holder         1   
358466   539988       84849B             fairy soap soap holder         1   
358467   539988        22854        cream sweetheart egg holder         2   
358468   539988       47559B                tea time oven glove         2   

                 order_date  price customer_id       date order_status  amount  
0       2010-01-04 09:43:00   4.25     14590.0 2010-01-04    cancelled    4.25  
1       2010-01-04 10:28:00   2.55     14590.0 2010-01-04    delivered   91.80  
2       2010-01-04 10:28:00   4.25     14590.0 2010-01-04    delivered   51.00  
3       2010-01-04 10:28:00   2.55     14590.0 2010-01-04    delivered    5.10  
4       2010-01-04 10:28:00   4.25     14590.0 2010-01-04    delivered    8.50  
...                     ...    ...         ...        ...          ...     ...  
358464  2010-12-23 16:06:00   1.25     18116.0 2010-12-23    delivered    1.25  
358465  2010-12-23 16:06:00   1.69     18116.0 2010-12-23    delivered    1.69  
358466  2010-12-23 16:06:00   1.69     18116.0 2010-12-23    delivered    1.69  
358467  2010-12-23 16:06:00   4.95     18116.0 2010-12-23    delivered    9.90  
358468  2010-12-23 16:06:00   1.25     18116.0 2010-12-23    delivered    2.50  

[358469 rows x 10 columns]

In [ ]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358469 entries, 0 to 358468
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      358469 non-null  object        
 1   product_code  358469 non-null  object        
 2   product_name  358469 non-null  object        
 3   quantity      358469 non-null  int64         
 4   order_date    358469 non-null  object        
 5   price         358469 non-null  float64       
 6   customer_id   358469 non-null  object        
 7   date          358469 non-null  datetime64[ns]
 8   order_status  358469 non-null  object        
 9   amount        358469 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(6)
memory usage: 27.3+ MB


# Membuat RFM segmentation

## Agregat data transaksi ke bentuk summary total transaksi (order), total nilai order (order value), tanggal order terakhir dari setiap pengguna

In [ ]:
df_user = df_clean.groupby('customer_id', as_index=False).agg(
    order_cnt=('order_id', 'nunique'),
    max_order_date =('order_date', 'max'),
    total_order_value = ('amount', 'sum'))

df_user

customer_id  order_cnt       max_order_date  total_order_value
0        12346.0          5  2010-10-04 09:54:00             602.40
1        12608.0          1  2010-10-31 10:49:00             415.79
2        12745.0          2  2010-08-10 10:14:00             723.85
3        12746.0          2  2010-06-30 08:19:00             266.35
4        12747.0         19  2010-12-13 10:41:00            4094.79
...          ...        ...                  ...                ...
3884     18283.0          6  2010-11-22 15:30:00             641.77
3885     18284.0          2  2010-10-06 12:31:00             486.68
3886     18285.0          1  2010-02-17 10:24:00             427.00
3887     18286.0          2  2010-08-20 11:57:00             941.48
3888     18287.0          4  2010-11-22 11:51:00            2345.71

[3889 rows x 4 columns]

In [ ]:
df_user['max_order_date'] = pd.to_datetime(df_user['max_order_date'])  # Convert 'max_order_date' to datetime
today = pd.to_datetime(df_clean['date'].max())  # Convert 'today' to datetime
df_user['day_since_last_order'] = (today - df_user['max_order_date']).dt.days


## Buat kolom jumlah hari sejak order terakhir

In [ ]:
today = df_clean['date'].max()
df_user['max_order_date'] = pd.to_datetime(df_user['max_order_date'])  # Convert 'max_order_date' to datetime
today = pd.to_datetime(df_clean['date'].max())  # Convert 'today' to datetime
df_user['day_since_last_order'] = (today - df_user['max_order_date']).dt.days

df_user

customer_id  order_cnt      max_order_date  total_order_value  \
0        12346.0          5 2010-10-04 09:54:00             602.40   
1        12608.0          1 2010-10-31 10:49:00             415.79   
2        12745.0          2 2010-08-10 10:14:00             723.85   
3        12746.0          2 2010-06-30 08:19:00             266.35   
4        12747.0         19 2010-12-13 10:41:00            4094.79   
...          ...        ...                 ...                ...   
3884     18283.0          6 2010-11-22 15:30:00             641.77   
3885     18284.0          2 2010-10-06 12:31:00             486.68   
3886     18285.0          1 2010-02-17 10:24:00             427.00   
3887     18286.0          2 2010-08-20 11:57:00             941.48   
3888     18287.0          4 2010-11-22 11:51:00            2345.71   

      day_since_last_order  
0                       79  
1                       52  
2                      134  
3                      175  
4                        9  
...                    ...  
3884                    30  
3885                    77  
3886                   308  
3887                   124  
3888                    30  

[3889 rows x 5 columns]

In [ ]:
df_user.describe()  # info kilas

order_cnt  total_order_value  day_since_last_order
count  3889.000000        3889.000000           3889.000000
mean      5.128568        1544.623084             89.800720
std       8.499330        3434.816315             88.873286
min       1.000000           1.250000             -1.000000
25%       1.000000         296.360000             24.000000
50%       3.000000         648.200000             57.000000
75%       6.000000        1585.940000            125.000000
max     163.000000       71970.390000            351.000000

## Buat binning dari jumlah hari sejak order terakhir yang terdiri dari 5 bins dengan batas-batasnya merupakan min, P20, P40, P60, P80, max dan beri label 1 sampai 5 dari bin tertinggi ke terendah sebagai skor recency

In [ ]:
df_user['recency_score'] = pd.cut(df_user['day_since_last_order'],
                                  bins = [df_user['day_since_last_order'].min(),
                                  np.percentile(df_user['day_since_last_order'], 20),
                                  np.percentile(df_user['day_since_last_order'], 40),
                                  np.percentile(df_user['day_since_last_order'], 60),
                                  np.percentile(df_user['day_since_last_order'], 80),
                                  df_user['day_since_last_order']. max()],
                                  labels=[5, 4, 3, 2, 1],
                                  include_lowest=True).astype(int)

df_user

customer_id  order_cnt      max_order_date  total_order_value  \
0        12346.0          5 2010-10-04 09:54:00             602.40   
1        12608.0          1 2010-10-31 10:49:00             415.79   
2        12745.0          2 2010-08-10 10:14:00             723.85   
3        12746.0          2 2010-06-30 08:19:00             266.35   
4        12747.0         19 2010-12-13 10:41:00            4094.79   
...          ...        ...                 ...                ...   
3884     18283.0          6 2010-11-22 15:30:00             641.77   
3885     18284.0          2 2010-10-06 12:31:00             486.68   
3886     18285.0          1 2010-02-17 10:24:00             427.00   
3887     18286.0          2 2010-08-20 11:57:00             941.48   
3888     18287.0          4 2010-11-22 11:51:00            2345.71   

      day_since_last_order  recency_score  
0                       79              2  
1                       52              3  
2                      134              2  
3                      175              1  
4                        9              5  
...                    ...            ...  
3884                    30              4  
3885                    77              2  
3886                   308              1  
3887                   124              2  
3888                    30              4  

[3889 rows x 6 columns]

## Buat binning dari total transaksi (order) yang terdiri dari 5 bins dengan batas-batasnya merupakan min, P20, P40, P60, P80, max dan beri label 1 sampai 5 dari bin terendah ke tertinggi sebagai skor frequency

In [ ]:
df_user['frequency_score'] = pd.cut(df_user['order_cnt'],
                                  bins = [0,
                                  np.percentile(df_user['order_cnt'], 20),
                                  np.percentile(df_user['order_cnt'], 40),
                                  np.percentile(df_user['order_cnt'], 60),
                                  np.percentile(df_user['order_cnt'], 80),
                                  df_user['order_cnt']. max()],
                                  labels=[1,2,3,4,5],
                                  include_lowest=True).astype(int)

df_user

customer_id  order_cnt      max_order_date  total_order_value  \
0        12346.0          5 2010-10-04 09:54:00             602.40   
1        12608.0          1 2010-10-31 10:49:00             415.79   
2        12745.0          2 2010-08-10 10:14:00             723.85   
3        12746.0          2 2010-06-30 08:19:00             266.35   
4        12747.0         19 2010-12-13 10:41:00            4094.79   
...          ...        ...                 ...                ...   
3884     18283.0          6 2010-11-22 15:30:00             641.77   
3885     18284.0          2 2010-10-06 12:31:00             486.68   
3886     18285.0          1 2010-02-17 10:24:00             427.00   
3887     18286.0          2 2010-08-20 11:57:00             941.48   
3888     18287.0          4 2010-11-22 11:51:00            2345.71   

      day_since_last_order  recency_score  frequency_score  
0                       79              2                4  
1                       52              3                1  
2                      134              2                2  
3                      175              1                2  
4                        9              5                5  
...                    ...            ...              ...  
3884                    30              4                4  
3885                    77              2                2  
3886                   308              1                1  
3887                   124              2                2  
3888                    30              4                3  

[3889 rows x 7 columns]

In [ ]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3889 entries, 0 to 3888
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   customer_id           3889 non-null   object        
 1   order_cnt             3889 non-null   int64         
 2   max_order_date        3889 non-null   datetime64[ns]
 3   total_order_value     3889 non-null   float64       
 4   day_since_last_order  3889 non-null   int64         
 5   recency_score         3889 non-null   int64         
 6   frequency_score       3889 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 212.8+ KB


## Buat kolom nama segmen berdasarkan skor recency dan frequency

In [ ]:
df_user['segment'] = np.select(
    [(df_user['recency_score']==5) & (df_user['frequency_score']>=4),
     (df_user['recency_score'].between(3, 4)) & (df_user['frequency_score']>=4),
     (df_user['recency_score']>=4) & (df_user['frequency_score'].between(2, 3)),
     (df_user['recency_score']<=2) & (df_user['frequency_score']==5),
     (df_user['recency_score']==3) & (df_user['frequency_score']==3),
     (df_user['recency_score']==5) & (df_user['frequency_score']==1),
     (df_user['recency_score']==4) & (df_user['frequency_score']==1),
     (df_user['recency_score']<=2) & (df_user['frequency_score'].between(3, 4)),
     (df_user['recency_score']==3) & (df_user['frequency_score']<=2),
     (df_user['recency_score']<=2) & (df_user['frequency_score']<=2)],
    ['01-Champion', '02-Loyal Customers', '03-Potential Loyalists', "04-Can't Lose Them", '05-Need Attention',
     '06-New Customers', '07-Promising', '08-At Risk', '09-About to Sleep', '10-Hibernating']
)
df_user

customer_id  order_cnt      max_order_date  total_order_value  \
0        12346.0          5 2010-10-04 09:54:00             602.40   
1        12608.0          1 2010-10-31 10:49:00             415.79   
2        12745.0          2 2010-08-10 10:14:00             723.85   
3        12746.0          2 2010-06-30 08:19:00             266.35   
4        12747.0         19 2010-12-13 10:41:00            4094.79   
...          ...        ...                 ...                ...   
3884     18283.0          6 2010-11-22 15:30:00             641.77   
3885     18284.0          2 2010-10-06 12:31:00             486.68   
3886     18285.0          1 2010-02-17 10:24:00             427.00   
3887     18286.0          2 2010-08-20 11:57:00             941.48   
3888     18287.0          4 2010-11-22 11:51:00            2345.71   

      day_since_last_order  recency_score  frequency_score  \
0                       79              2                4   
1                       52              3                1   
2                      134              2                2   
3                      175              1                2   
4                        9              5                5   
...                    ...            ...              ...   
3884                    30              4                4   
3885                    77              2                2   
3886                   308              1                1   
3887                   124              2                2   
3888                    30              4                3   

                     segment  
0                 08-At Risk  
1          09-About to Sleep  
2             10-Hibernating  
3             10-Hibernating  
4                01-Champion  
...                      ...  
3884      02-Loyal Customers  
3885          10-Hibernating  
3886          10-Hibernating  
3887          10-Hibernating  
3888  03-Potential Loyalists  

[3889 rows x 8 columns]

In [ ]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3889 entries, 0 to 3888
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   customer_id           3889 non-null   object        
 1   order_cnt             3889 non-null   int64         
 2   max_order_date        3889 non-null   datetime64[ns]
 3   total_order_value     3889 non-null   float64       
 4   day_since_last_order  3889 non-null   int64         
 5   recency_score         3889 non-null   int64         
 6   frequency_score       3889 non-null   int64         
 7   segment               3889 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(2)
memory usage: 243.2+ KB


## Tampilkan summary dari RFM segmentation (poin 8) berupa banyaknya pengguna, rata-rata dan median dari total order, total order value, dan jumlah hari sejak order terakhir

In [ ]:
summary = pd.pivot_table(df_user, index= 'segment',
                         values = ['customer_id','day_since_last_order','order_cnt','total_order_value'],
                          aggfunc={'customer_id': pd.Series.nunique,
                        'day_since_last_order': [np.mean, np.median],
                        'order_cnt': [np.mean, np.median],
                        'total_order_value': [np.mean, np.median]})
summary['pct_unique'] = (summary['customer_id'] / summary['customer_id'].sum() * 100).round(1)
summary

customer_id day_since_last_order         order_cnt  \
                           nunique                 mean median       mean   
segment                                                                     
01-Champion                    550             9.618182    8.5  15.467273   
02-Loyal Customers             546            39.864469   36.0   8.767399   
03-Potential Loyalists         523            22.573614   23.0   2.829828   
04-Can't Lose Them              64           120.984375  111.5  11.375000   
05-Need Attention              176            57.613636   58.0   3.397727   
06-New Customers                50            13.220000   15.0   1.000000   
07-Promising                   142            31.760563   33.0   1.000000   
08-At Risk                     426           139.455399  119.0   4.136150   
09-About to Sleep              352            57.735795   57.0   1.417614   
10-Hibernating                1060           195.837736  198.0   1.313208   

                              total_order_value           pct_unique  
                       median              mean    median             
segment                                                               
01-Champion              10.0       5003.674245  2775.525       14.1  
02-Loyal Customers        7.0       2622.817826  1946.850       14.0  
03-Potential Loyalists    3.0        766.769828   622.070       13.4  
04-Can't Lose Them        9.5       2839.948125  2268.405        1.6  
05-Need Attention         3.0        989.232676   826.370        4.5  
06-New Customers          1.0        244.689000   193.675        1.3  
07-Promising              1.0        287.800282   238.440        3.7  
08-At Risk                4.0       1153.825683   875.430       11.0  
09-About to Sleep         1.0        448.229688   334.755        9.1  
10-Hibernating            1.0        343.083842   257.005       27.3

# Analisa Data

1. Pengguna paling banyak berada pada segmen Hibernating (1060 atau 27.3%), Champion (550 atau 14.1%), dan Loyal Customers (546 atau 14.0%).

2. Program khusus yang fokus pada urgensi bertransaksi untuk Loyal Customers (546 atau 14.0%) dapat dibuat untuk membuat mereka bertransaksi kembali dalam waktu dekat sehingga bisa naik ke segmen Champion.

3. Program khusus yang fokus pada jumlah transaksi untuk Potential Loyalists (523 atau 13.4%) dapat dibuat untuk membuat mereka lebih sering bertransaksi sehingga bisa naik ke segmen Champion

4. Program khusus untuk Hibernating (1060 atau 27.3%) dapat dibuat untuk membuat mereka kembali bertransaksi walaupun belum begitu sering sehingga bisa naik ke segmen New Customers atau bahkan Potential Loyalists.

# Thank You